### ISSUE TITLE
Unable to gather system info et al on CentOS 7 clients

### ISSUE THREAD

**Issue by pirxthepilot** [zentralopensource/zentral#53]
I'm using v0.4.0 of Zentral and am able to add CentOS 6 and 7 hosts, as well as run queries, etc. The problem is that when viewing CentOS 7 hosts, the section that is supposed to show system info (CPU, Memory, OS version, networking) is empty.

<img width="834" alt="screen shot 2017-09-29 at 4 37 37 pm" src="https://user-images.githubusercontent.com/7690118/31039739-029a2b1a-a535-11e7-83e7-1a344080b945.png">

I think this may be the relevant error from the logs:

```
I0929 16:32:39.288528 17443 distributed.cpp:133] Executing distributed query: __zentral_distributed_inventory_query_os_version: select 'os_version' as table_name, name, major, minor, patch, build from os_version;
I0929 16:32:39.288874 17443 virtual_table.cpp:291] Error casting minor () to INTEGER
I0929 16:32:39.288892 17443 virtual_table.cpp:291] Error casting patch () to INTEGER
```

**Comment by np5** [comment]
It looks like an osquery issue. Did you try to query the `os_version` table ? with `osqueryi` directly on the clients ? The query we run for the inventory is [here](https://github.com/zentralopensource/zentral/blob/8165b2b0cc3fcfd0faee8bf854d15104f829b1be/zentral/contrib/osquery/conf.py#L9).

**Comment by pirxthepilot** [comment]
The table returns empty values:

```
+------------+--------------+-------+-------+-------+-------+
| table_name | name         | major | minor | patch | build |
+------------+--------------+-------+-------+-------+-------+
| os_version | CentOS Linux | 7     |       |       |       |
+------------+--------------+-------+-------+-------+-------+
```

On CentOS 6 minor and patch are populated:

```
+------------+--------+-------+-------+-------+-------+
| table_name | name   | major | minor | patch | build |
+------------+--------+-------+-------+-------+-------+
| os_version | CentOS | 6     | 6     | 0     |       |
+------------+--------+-------+-------+-------+-------+
```

BTW, the osquery version is 2.8.0.

**Comment by pirxthepilot** [comment]
Looked at the code for a bit - could [this](https://github.com/zentralopensource/zentral/blob/master/zentral/contrib/inventory/clients/watchman.py#L184) be the one causing the casting errors? If so, maybe I can insert a quick try-catch into it?

**Comment by np5** [comment]
The errors in the logs you posted clearly show an error in a [C++ source file](https://github.com/facebook/osquery/blob/9e779a56e607ade210e06f15a019436e9f732157/osquery/sql/virtual_table.cpp#L291). This is an osquery issue, on the client side.

The code you linked too, runs on the server, and is used to get the inventory information of the machines present in a [Watchman](https://www.watchmanmonitoring.com/) account. It is not related to the present issue.

**Comment by pirxthepilot** [comment]
Oh, gotcha. Created an [issue](https://github.com/facebook/osquery/issues/3753) on the osquery github. 

Can you confirm if this error is actually what causes zentral not to display system information?

Thanks for the help!

**Comment by np5** [comment]
The [minor version number](https://github.com/zentralopensource/zentral/blob/8165b2b0cc3fcfd0faee8bf854d15104f829b1be/zentral/contrib/inventory/models.py#L213) of the OS is mandatory in the Django model. That could indeed cause an issue if it is not present. What do you have in your Django logs ?

**Comment by np5** [comment]
Try with 9ce499c. It should work.

**Comment by pirxthepilot** [comment]
@np5 Thanks! I tried the commit but I'm getting a 404 from nginx this time.

**Comment by pirxthepilot** [comment]
So I applied the change in 3d6bfa1f5b21908bb1735dc34b83ed87c2693338 against the v0.4.0 branch, and it runs now, but unfortunately the issue isn't fixed. Django log:

```
web_1       | 2017-10-02 18:49:48,327 PID14 utils ERROR Could not commit machine snapshot
web_1       | Traceback (most recent call last):
web_1       |   File "/zentral/zentral/utils/mt_models.py", line 102, in commit
web_1       |     obj = self.get(mt_hash=tree['mt_hash'])
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
web_1       |     return getattr(self.get_queryset(), name)(*args, **kwargs)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 385, in get
web_1       |     self.model._meta.object_name
web_1       | zentral.contrib.inventory.models.DoesNotExist: MachineSnapshot matching query does not exist.
web_1       |
web_1       | During handling of the above exception, another exception occurred:
web_1       |
web_1       | Traceback (most recent call last):
web_1       |   File "/zentral/zentral/utils/mt_models.py", line 102, in commit
web_1       |     obj = self.get(mt_hash=tree['mt_hash'])
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
web_1       |     return getattr(self.get_queryset(), name)(*args, **kwargs)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 385, in get
web_1       |     self.model._meta.object_name
web_1       | zentral.contrib.inventory.models.DoesNotExist: OSVersion matching query does not exist.
web_1       |
web_1       | During handling of the above exception, another exception occurred:
web_1       |
web_1       | Traceback (most recent call last):
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
web_1       |     return self.cursor.execute(sql, params)
web_1       | psycopg2.IntegrityError: null value in column "minor" violates not-null constraint
web_1       | DETAIL:  Failing row contains (242, 546d4101251ad7625ee1896990a2107f45b808a9, 2017-10-02 18:49:48.325612+00, CentOS Linux, 7, null, null, null).
web_1       |
web_1       |
web_1       | The above exception was the direct cause of the following exception:
web_1       |
web_1       | Traceback (most recent call last):
web_1       |   File "/zentral/zentral/contrib/inventory/utils.py", line 130, in commit_machine_snapshot_and_trigger_events
web_1       |     machine_snapshot_commit, machine_snapshot = MachineSnapshotCommit.objects.commit_machine_snapshot_tree(tree)
web_1       |   File "/zentral/zentral/contrib/inventory/models.py", line 403, in commit_machine_snapshot_tree
web_1       |     machine_snapshot, _ = MachineSnapshot.objects.commit(tree)
web_1       |   File "/zentral/zentral/utils/mt_models.py", line 122, in commit
web_1       |     fk_obj, _ = f.related_model.objects.commit(v)
web_1       |   File "/zentral/zentral/utils/mt_models.py", line 145, in commit
web_1       |     raise integrity_error
web_1       |   File "/zentral/zentral/utils/mt_models.py", line 136, in commit
web_1       |     obj.save(**extra_obj_save_kwargs)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 796, in save
web_1       |     force_update=force_update, update_fields=update_fields)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 824, in save_base
web_1       |     updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 908, in _save_table
web_1       |     result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 947, in _do_insert
web_1       |     using=using, raw=raw)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
web_1       |     return getattr(self.get_queryset(), name)(*args, **kwargs)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 1045, in _insert
web_1       |     return query.get_compiler(using=using).execute_sql(return_id)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1054, in execute_sql
web_1       |     cursor.execute(sql, params)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
web_1       |     return super(CursorDebugWrapper, self).execute(sql, params)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
web_1       |     return self.cursor.execute(sql, params)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
web_1       |     six.reraise(dj_exc_type, dj_exc_value, traceback)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
web_1       |     raise value.with_traceback(tb)
web_1       |   File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
web_1       |     return self.cursor.execute(sql, params)
web_1       | django.db.utils.IntegrityError: null value in column "minor" violates not-null constraint
web_1       | DETAIL:  Failing row contains (242, 546d4101251ad7625ee1896990a2107f45b808a9, 2017-10-02 18:49:48.325612+00, CentOS Linux, 7, null, null, null).
```

**Comment by np5** [comment]
Did you run the django db migrations ?

**Comment by pirxthepilot** [comment]
Sorry, I'm quite new to this. How do I do db migration through docker?

**Comment by np5** [comment]
If you are using the docker-compose configuration of the repository, you just have to restart the container.

Manually:

```
docker-compose run --rm web python server/manage.py migrate
```

**Comment by pirxthepilot** [comment]
In that case, I have already restarted with docker-compose.

**Comment by np5** [comment]
Could you connect to the postgres db, dump the content of the `django_migrations` table and the structure of the `inventory_osversion` table?

With the current `docker-compose` setup:

To get the content of the `django_migrations` table as CSV on the `stdout`:
```
docker run -it --network zentral_default --rm postgres:9.6 \
       sh -c "exec psql -U zentral -h zentral_db_1 zentral -c 'COPY (SELECT * FROM django_migrations) TO STDOUT WITH CSV'"
```
The [password](https://github.com/zentralopensource/zentral/blob/4e1f748e4259f707d1f26d1391881cd69a2f954b/conf/start/docker/postgres.env#L2) is `zentral`.

To get the structure of the `inventory_osversion`table:
```
docker run -it --network zentral_default --rm postgres:9.6 \
       sh -c "exec psql -U zentral -h zentral_db_1 zentral -c '\d inventory_osversion'"
```

**Comment by np5** [comment]
See e29ac133 for an improved enrollment script that works with CentOS as well.

**Comment by pirxthepilot** [comment]
Awesome! With the latest build I am able to get CentOS 7 info now, and my clients are now classified as Linux. However, I can't seem to view events on any host now - not even the CentOS 6 ones. Wonder if you're getting the same behavior?

(Note: I did a complete rebuild with docker-compose)

**Comment by np5** [comment]
If the CentOS machines show up in the inventory, it means that the communication with Zentral is working (which is in contradiction with the log lines you included. Are they old ?).

The events are stored in Elasticsearch by the `workers` container. You can check the status of the workers in the `Targets` tab of Prometheus (accessible via the `extra links` menu in Zentral).

I don't really know what kind of docker setup you are running, so I can only guess. Did you update the `workers` container as well ? The Zentral source code of this container must match the state of the database structure.

**Comment by pirxthepilot** [comment]
I rebuilt the workers container and am now getting events for heartbeat, distributed_read and config. Still no logs from probes, though.. Checked Prometheus targets and everything is online.

```
Oct 04 13:34:19 osquery6 osqueryd[27023]: I1004 13:34:19.678167 27047 scheduler.cpp:75] Executing scheduled query listening-ports_cdb24fa9: select * from listening_ports;
Oct 04 13:34:30 osquery6 osqueryd[27023]: I1004 13:34:30.183815 27036 config.cpp:939] Refreshing configuration state
Oct 04 13:34:30 osquery6 osqueryd[27023]: I1004 13:34:30.184058 27036 tls.cpp:205] TLS/HTTPS POST request to URI: https://zentral/osquery/config
Oct 04 13:34:30 osquery6 osqueryd[27023]: I1004 13:34:30.439059 27039 tls.cpp:205] TLS/HTTPS POST request to URI: https://zentral/osquery/log
Oct 04 13:34:30 osquery6 osqueryd[27023]: I1004 13:34:30.464298 27039 buffered.cpp:72] Error sending results to logger: JSON deserialize error: <unspecified file>(1): expected value
```

The way I understand it, osquery is able to fetch the query from /distributed_read, but errors out when trying to post the result to /log.

**Comment by pirxthepilot** [comment]
Update: I switched back to v0.4.0 and posting to /log is working.

**Comment by np5** [comment]
Could you try to find the errors in the logs and post them in this thread please ?

```
docker-compose logs web
```

**Comment by pirxthepilot** [comment]
Ok, looks like the `osquery distributed query probe` is good (which I believe posts to /distributed/write?). `osquery probe` and `osquery fim probe` both fail.

These are the only logs I found that are relevant:

```
2017-10-04 22:10:54,987 PID29 api_views ERROR APIAuthError osquery reported SN VMware-56 4d 2a d7 59 f7 9d 5a-d4 4c c5 24 46 c1 cb 6e different from enrollment SN 564D2AD7-59F7-9D5A-D44C-C52446C1CB6D
2017-10-04 22:10:54,988 PID29 basehttp WARNING "POST /osquery/log HTTP/1.0" 403 140
```

**Comment by np5** [comment]
Nice catch. Fixed with 19c59c7.

With [198a6159](https://github.com/zentralopensource/zentral/commit/198a61593c8545a5fecfcc173edab5de9240c4ca#diff-15485124fda604883c9f71cd63a83af2R224), we added a check to detect the serial number conflicts. Some of our macOS machines were cloned, and the enrollment secret was the same on multiple machines. 

This introduced a problem with Linux clients, because the serial numbers built by `osquery` are DMI/SMBIOS based, and the ones we generate in the enrollment script are different (with 2546aa2 even more). This is why the log requests were rejected.

We need more tests!

**Comment by pirxthepilot** [comment]
Works great now! Thanks so much!

Since my original issue is solved (more than solved, actually!) I will go ahead and close this issue.

I'll keep testing zentral (we have a mix of CentOS and Scientific Linux hosts) and keep reporting issues if I find more :)
